### PR TITLE
TSK-1642: now resolving complex objects during change detection properly

### PR DESCRIPTION
**Pull request by mustaphazorgati**
https://sonarcloud.io/dashboard?branch=task%2F1642-2&id=mustaphazorgati_taskana

---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [ ] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java [hunk 1]
```diff
@@ -26,51 +28,22 @@ private ObjectAttributeChangeDetector() {}
    * @throws SystemException when any parameter is null
    */
   public static <T> String determineChangesInAttributes(T oldObject, T newObject) {
-    List<Field> fields = new ArrayList<>();
-
-    if (Objects.isNull(oldObject) || Objects.isNull(newObject)) {
+    if (oldObject == null || newObject == null) {
       throw new SystemException(
           "Null was provided as a parameter. Please provide two objects of the same type");
     }
```
Comment (mustaphazorgati) [inline common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java:34]: should  we additionally check that oldObject and newObject are really the same class? The bound type parameter `T` only verifies that they are compatible to each other. oldObject could be a subclass of newObject. Currently we would generate more changes than they really might be..

File: common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java [hunk 2]
```diff
@@ -79,26 +52,42 @@ private ObjectAttributeChangeDetector() {}
...
-      fields.addAll(Arrays.asList(currentClass.getDeclaredFields()));
-      currentClass = currentClass.getSuperclass();
+  private static JSONObject generateChangedAttribute(
+      Field field, Object oldValue, Object newValue) {
+    JSONObject changedAttribute = new JSONObject();
+    changedAttribute.put("fieldName", field.getName());
+    changedAttribute.put(
+        "oldValue", Optional.ofNullable(oldValue).map(JSONObject::wrap).orElse(""));
+    changedAttribute.put(
+        "newValue", Optional.ofNullable(newValue).map(JSONObject::wrap).orElse(""));
```
Comment (mustaphazorgati) [inline common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java:62]: @gitgoodjhe: The JSONObject#wrap is the "effective" change. Everything else is refactoring

File: common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java (hunk 1 above)
Comment (gitgoodjhe) [reply common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java:34]: That's right, even though I would expect the dev who uses this to provide two objects of the same class. Also it's debatable if we would actually want to compare subclasses also. If that's the case, then generating the changes wouldn't be an error in my opinion.

File: common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java [hunk 3]
```diff
@@ -26,51 +28,22 @@ private ObjectAttributeChangeDetector() {}
...
-                      Optional.ofNullable(fieldAndValuePairTriplet.getMiddle()).orElse(""));
-                  changedAttribute.put(
-                      "newValue",
-                      Optional.ofNullable(fieldAndValuePairTriplet.getRight()).orElse(""));
-                  return changedAttribute;
-                })
+            .filter(field -> !"customAttributes".equals(field.getName()))
+            .map(wrap(field -> Triplet.of(field, field.get(oldObject), field.get(newObject))))
+            .filter(t -> !Objects.equals(t.getMiddle(), t.getRight()))
+            .map(t -> generateChangedAttribute(t.getLeft(), t.getMiddle(), t.getRight()))
```
Comment (gitgoodjhe) [inline common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java:46]: Nice, I like putting this in a separate method

File: common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java [hunk 4]
```diff
@@ -79,26 +52,42 @@ private ObjectAttributeChangeDetector() {}
...
+    JSONObject changedAttribute = new JSONObject();
+    changedAttribute.put("fieldName", field.getName());
+    changedAttribute.put(
+        "oldValue", Optional.ofNullable(oldValue).map(JSONObject::wrap).orElse(""));
+    changedAttribute.put(
+        "newValue", Optional.ofNullable(newValue).map(JSONObject::wrap).orElse(""));
+    return changedAttribute;
+  }
+
+  private static Stream<Field> retrieveAllFields(Class<?> clazz) {
```
Comment (gitgoodjhe) [inline common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java:66]: this went from 3 LOC to 6 LOC and for me was more readable before. What's the advantage here?

**Review by gitgoodjhe — APPROVED** [review]

File: common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java (hunk 1 above)
Comment (mustaphazorgati) [reply common/taskana-common/src/main/java/pro/taskana/common/internal/util/ObjectAttributeChangeDetector.java:34]: alright. Then we keep it like this ;)

**Review by gitgoodjhe — APPROVED** [review]

**Pull request merged**